### PR TITLE
What is new with OAuth 2 1 post

### PR DESCRIPTION
**Pull request by mooreds**
Added a post about what is new with OAuth 2.1.

I tried to make the tone a bit less formal.

Not sure it makes sense to add in more URLs about the redirect attacks, as the one you suggested is pretty good.

I added in an image of the admin configuration, but it is from the Angular post--can someone show me how to create a unique one? Or if that's a hassle, I am ok with leaving it this way.

I will need a custom header image--how are those created?

### CODE REVIEW

File: site/_posts/2020-04-15-whats-new-in-oauth-2-1.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+[Refresh tokens](https://tools.ietf.org/html/rfc6749#section-6) allow a client to retrieve new access tokens without reauthentication. This is helpful if you need access to a resource for longer than an access token live, or if you need infrequent access (such as being logged into your email for months or years). As such, they are typically longer lived than access tokens. Therefore you should take twice as much care when it comes to securing a refresh token. 
+
+If they are acquired by an attacker, the attacker can create access tokens at will. Obviously at that point the resource which the access tokens protect will no longer be secure. As an example of how to secure refresh tokens, here’s a [post using the Authorization Code grant](https://fusionauth.io/learn/expert-advice/authentication/spa/oauth-authorization-code-grant-jwts-refresh-tokens-cookies) which stores refresh tokens securely using HttpOnly cookies and a limited cookie domain. Incidentally, never share your refresh tokens between different devices.
+
+The OAuth 2.1 draft specification provides two options for refresh tokens: they can be one time use or tied to the sender with a cryptographic binding.
```
Comment (pgeoly) [inline site/_posts/2020-04-15-whats-new-in-oauth-2-1.md:137]: Hyphenate One-time
```suggestion
The OAuth 2.1 draft specification provides two options for refresh tokens: they can be one-time use or tied to the sender with a cryptographic binding. 
```

File: site/_posts/2020-04-15-whats-new-in-oauth-2-1.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+The OAuth 2.1 draft specification provides two options for refresh tokens: they can be one time use or tied to the sender with a cryptographic binding. 
+
+One time use means that after a refresh token (call it refresh token A) is used to retrieve an access token, it becomes invalid. Of course, the OAuth server can send a new refresh token (call it refresh token B) along with the requested access token. In this case, once the newly delivered access token expires, the client can request another access token using refresh token B, receive the new access token and a new refresh token C, and so on and so on. The change to one time use refresh tokens may require changing client code to store the new refresh token on every refresh.
```
Comment (pgeoly) [inline site/_posts/2020-04-15-whats-new-in-oauth-2-1.md:139]: Hyphenate One-time
```suggestion
One time use means that after a refresh token (call it refresh token A) is used to retrieve an access token, it becomes invalid. Of course, the OAuth server can send a new refresh token (call it refresh token B) along with the requested access token. In this case, once the newly delivered access token expires, the client can request another access token using refresh token B, receive the new access token and a new refresh token C, and so on and so on. The change to one-time use refresh tokens may require changing client code to store the new refresh token on every refresh.
```

File: site/_posts/2020-04-15-whats-new-in-oauth-2-1.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+The other recommended option is to ensure the OAuth server cryptographically binds the refresh token to the client. The options mentioned in the the "OAuth 2.0 Security Best Current Practices" document are [OAuth token binding](https://www.ietf.org/archive/id/draft-ietf-oauth-token-binding-08.txt) or Mutual TLS authentication [RFC 8705](https://tools.ietf.org/html/rfc8705). This binding ensures the request came from the client to which the refresh token was issued.
```
Comment (pgeoly) [inline site/_posts/2020-04-15-whats-new-in-oauth-2-1.md:141]:
```suggestion
The other recommended option is to ensure the OAuth server cryptographically binds the refresh token to the client. The options mentioned in the "OAuth 2.0 Security Best Current Practices" document are [OAuth token binding](https://www.ietf.org/archive/id/draft-ietf-oauth-token-binding-08.txt) or Mutual TLS authentication [RFC 8705](https://tools.ietf.org/html/rfc8705). This binding ensures the request came from the client to which the refresh token was issued. 
```

File: site/_posts/2020-04-15-whats-new-in-oauth-2-1.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+The other recommended option is to ensure the OAuth server cryptographically binds the refresh token to the client. The options mentioned in the the "OAuth 2.0 Security Best Current Practices" document are [OAuth token binding](https://www.ietf.org/archive/id/draft-ietf-oauth-token-binding-08.txt) or Mutual TLS authentication [RFC 8705](https://tools.ietf.org/html/rfc8705). This binding ensures the request came from the client to which the refresh token was issued. 
+
+To sum up, the OAuth 2.1 spec requires an OAuth server protect refresh tokens by requiring them to either be one time use or by using cryptographic proof of client association.
```
Comment (pgeoly) [inline site/_posts/2020-04-15-whats-new-in-oauth-2-1.md:143]: Add hyphen to one-time
```suggestion
To sum up, the OAuth 2.1 spec requires an OAuth server protect refresh tokens by requiring them to either be one-time use or by using cryptographic proof of client association. 
```

File: site/_posts/2020-04-15-whats-new-in-oauth-2-1.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+When writing a client application, avoid the Implicit grant and the Resource Owner Password Credentials grant. However, as these are part of the OAuth 2.0 specification, they are currently supported by FusionAuth.
+
+Make sure your OAuth server is doing to following:
+* Using PKCE whenever you use the Authorization Code grant. ([FusionAuth highly recommends using PKCE.](https://fusionauth.io/docs/v1/tech/oauth/#example-authorization-code-grant))
+* Make sure that your redirect URIs are compared using exact string matches, not wildcard or substring matching. (FusionAuth has you covered: ["URLs that are not authorized [configured in the application] may not be utilized in the redirect_uri"](https://fusionauth.io/docs/v1/tech/core-concepts/applications#oauth).)
+* Make sure bearer tokens are never present in the query string. (FusionAuth doesn’t support access tokens in the query string other than in the Implict Grant. But you shouldn’t be using that anyway.)
```
Comment (pgeoly) [inline site/_posts/2020-04-15-whats-new-in-oauth-2-1.md:158]: Corrected spelling - Implicit
```suggestion
* Make sure bearer tokens are never present in the query string. (FusionAuth doesn’t support access tokens in the query string other than in the Implicit Grant. But you shouldn’t be using that anyway.)
```

File: site/_posts/2020-04-15-whats-new-in-oauth-2-1.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+> Bearer token usage omits the use of bearer tokens in the query string of URIs as per Section 4.3.2 of [OAuth 2.0 Security Best Current Practices](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-14)
+
+> Refresh tokens must either be sender-constrained or one-time use as per Section 4.12.2 of [OAuth 2.0 Security Best Current Practices](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-14)
+
+Whew, that's a lot of jargon. We'll examine each of these in turn. But before we do, let’s define a few terms that will be used in the rest of this post.
+
+* A `client` is a piece of code that the user is interacting with; browsers, native apps or single page applications are all clients. 
+* An `OAuth server` implements OAuth specifications and has or can obtain information about which resources are available to clients--in the RFCs this is called an Authorization Server, but this is also known as an Identity Provider. Most users call it "the place I login".
```
[the remark this answers is not in the document]
Comment (mooreds) [reply site/_posts/2020-04-15-whats-new-in-oauth-2-1.md:60]: Totally the verb, thanks for catching that.

**Pull request merged**